### ISSUE TITLE
reCaptcha (v3) marking everything as spam

### ISSUE THREAD

**Issue by medoingthings** [verbb/formie#1274]
### Describe the bug

I want to use reCaptcha v3 to save me from spam. The form submits data using the HubSpot integration. That works great, as long as I don’t activate reCaptcha.

As soon as I activate the reCaptcha v3 integration, everything gets marked as spam.
Logs say: `Submission marked as spam - “Failed Captcha “Recaptcha”: “”” - {"email": …`

When I replace reCaptcha v3 with reCaptcha v2 (Checkbox), then it works. For the records: Whenever I switch between v2 and v3 I set the proper reCaptcha keys accordingly. So the config should be fine. I tried it with the minimum score set to 0.2, but I’m not even sure if it’s the score or something else, because it’s hard to debug with the above mentioned error message.

Any hint is much appreciated. I just recently switched to Formie and am completely blown away by the functionality and customizability. Amazing how much thought and experience went into this!

Cheers,
Mark

### Steps to reproduce

1. Setup form with HubSpot integration (not sure if HubSpot makes a difference here)
2. Setup reCaptcha v3 integration
3. Try to send form

### Form settings

- Multi-page form: No
- Submission Method: Page Reload (plus redirect to same page)
- Client-side Validation: No
- Custom Form Templates: No

### Craft CMS version

4.3.6.1

### Plugin version

2.0.21

### Multi-site?

Yes

### Additional context

Happens on production and locally (both domains are whitelisted in reCaptcha config).
As stated, I use the HubSpot integration to write the submission data to HubSpot using the form api.

**Comment by engram-design** [comment]
Yeah, unfortunately, the reCaptcha service doesn't give us much feedback on _what_ went wrong, which leads to unhelpful debug messages like that.

Using HubSpot shouldn't be an issue, as that's fired after the recaptcha is triggered and evaluated.

I certainly can't replicate that on my end, which is difficult. I've just added some more debugging to the integration, to hopefully try and get some more context on what's going on. To get this early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-4 as 2.0.21",
  "...": "..."
}
```

Then run `composer update`.

**Comment by medoingthings** [comment]
That was helpful in order to get closer 👍

The error now says `Client-side token missing`. Which is a bit weird, because I obviously set up the tokens (read them from `.env`).

I got suspicious if the HubSpot form might have a captcha itself which might have explained the behavior. But reCaptcha is not active on the HubSpot form.

This is the log – any ideas?:
```
2023-02-06 10:51:35 [INFO] Submission triggered for newsletterForm.
2023-02-06 10:51:35 [INFO] Submission marked as spam - “Failed Captcha “Recaptcha”: “Client-side token missing.”” - {"email":"mail@address.com","fullName":"{\"prefix\":null,\"firstName\":\"Mark\",\"middleName\":null,\"lastName\":null,\"name\":null,\"isMultiple\":true}","datenschutzZustimmung":true}.
2023-02-06 10:51:35 [ERROR] Couldn’t save submission due to errors - [].
2023-02-06 10:51:35 [INFO] Request context:
$_GET = [
    'p' => 'mentoring-programm'
]

$_POST = [
    'CRAFT_CSRF_TOKEN' => '••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••••'
    'action' => 'formie/submissions/submit'
    'submitAction' => 'submit'
    'handle' => 'newsletterForm'
    'siteId' => '1'
    'redirect' => 'e888a5zaab9e9ad25ef878a87c945da3d49301af795716c09438db099fe6593a/mentoring-programm#form-danke-213915'
    'fields' => [
        'email' => 'mail@address.com'
        'fullName' => [
            'firstName' => 'Mark'
            'lastName' => ''
        ]
        'datenschutzZustimmung' => '1'
    ]
    'g-recaptcha-response' => ''
]
```

**Comment by engram-design** [comment]
So the `Client-side token missing` error means that it's missing from the _request_ which you can see `g-recaptcha-response` is blank - that's the culprit.

Are there any JS errors on the front-end when you fill out the form? Using Recaptcha v3 should ping a request off to their server as soon as you hit the submit button.

There's a few things that could be the issue:
- A JS error is stopping the request/response to Recaptcha to generate the token
- Your front-end might use Vue (and similar) which needs its own handling
- The `formie.js` file isn't being loaded at all, meaning _no_ JS is being loaded. You can confirm this in your Network tab when you load the page. If that's the case (it's missing), there's something wrong with how you're rendering the form.

If you can confirm any of the above points?

**Comment by medoingthings** [comment]
Getting really close… first to your questions:
* No JS errors
* No VueJS involved
* `formie.js` and reCaptcha `api.js` are both being loaded properly

**But:**
I’ve setup a Form Template for my form where I disabled everything but `Output Base JavaScript`. When I use this template, reCaptcha doesn’t work. But as soon as I set the form template to the "Default Formie Template", it works.

I compared the resulting html files and there are only a couple of differences, but the one that stands out is the encoded JavaScript. I think it’s weird that the encoded JavaScript block is bigger in the "Basic JavaScript" file. I’ll send you both files via email. Hopefully you can get something out of it.

Cheers,
Mark

**Comment by engram-design** [comment]
Ah, that'll do it! You'll need the "Theme JS" in order to have everything work like client-side validation, captchas, ajax-submissons and lots more. If you disable that, you'll need to roll your own handling of _a lot_ of things, so it's not the most recommended approach, but it's there if you want to make yourself a lot of extra work!

So yep, you'll need to enable the Theme JS in order for captchas to work.

**Comment by medoingthings** [comment]
Oh 😅

As an improvement I would suggest to change the wording then. Because now it sounds like only the basic JS was mandatory. I don’t use the client side validation or any theming, so I thought I wouldn’t need the Theme JS.

Thanks for your support! Much appreciated.

<img width="691" alt="js-options" src="https://user-images.githubusercontent.com/1577712/217470498-53ef104f-a9e9-4b8f-8742-f4105ae51041.png">

**Comment by engram-design** [comment]
Ah, I stand corrected. You **don't** need the Theme JS to have captchas work, but there is indeed a bug where it's not working correctly without it.

Should be fixed with another `composer update`, or of course enabling the Theme JS will also do the trick. But again, just be mindful of all the things that aren't included with the Theme JS, like ajax-submission, validation and a few other things.

**Comment by medoingthings** [comment]
Yup, that seems to solve it. Thanks! Guess this will find its way into the next release, so closing the issue.

**Comment by medoingthings** [comment]
The issue seems to be back. 

With reCaptcha v2 set up and "Output Theme JS" off, I submit the form, but it only validates the Captcha. The form is not submitted.

As soon as I activate "Output Theme JS", the form works as expected.

Formie version 2.0.31

**Comment by engram-design** [comment]
Looks to be specific to the V2 Checkbox on this one, as it acts a little differently to the others.

Fixed for the next release. To get this early, run `composer require verbb/formie:"dev-craft-4 as 2.0.31"`.

**Comment by engram-design** [comment]
Fixed in [2.0.32](https://github.com/verbb/formie/releases/tag/2.0.32)
